odoo/addons/base/models/ir_mail_server.py

```diff
@@ -328,8 +328,17 @@ def build_email(self, email_from, email_to, subject, body, email_cc=None, email_
         msg['Message-Id'] = encode_header(message_id)
         if references:
             msg['references'] = encode_header(references)
+
+        force_smtp_from = self.env['ir.config_parameter'].sudo().get_param('mail.force.smtp.from')
+        email_from = encode_rfc2822_address_header(email_from)
+        if force_smtp_from:
+            msg['From'] = self._encapsulate_email(email_from, force_smtp_from)
```

Comment:
I think it would be useful to also extract the email From rewrite logic into a separate method (including the ICP checks), as that may become a feature that people want to override for more custom behavior.

Reply:
Moved to `_set_email_from`